### ISSUE TITLE
NullReferenceException wenn calling ToString on None

### ISSUE THREAD

**Issue by forki** [Microsoft/visualfsharp#4345]
None.ToString();;

leads to NullReferenceException in fsi and fsc.

#### Expected behavior

printing "<null>" like `printfn "%O" None` doesn

#### Actual behavior

NullReferenceException

#### Known workarounds

None

**Comment by matthid** [comment]
Isn't this kind of by design?

**Comment by AviAvni** [comment]
Yes it's by design look https://github.com/Microsoft/visualfsharp/pull/1776
but there is a way to handle this differently

**Comment by forki** [comment]
> but there is a way to handle this differently

what's that meaning?

**Comment by AviAvni** [comment]
the compiler can emit something else then 
``
IL_0000: ldnull
IL_0001: callvirt instance string [System.Runtime]System.Object::ToString()
```

maybe ignore the optimization https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core/prim-types.fs#L2971

or raise compile error

**Comment by AviAvni** [comment]
this is not only to None it's also to unit and even for custom discriminated union with this attribute 
```
[<CompilationRepresentation(CompilationRepresentationFlags.UseNullAsTrueValue)>]
```

**Comment by forki** [comment]
please no compile error ;-)

**Comment by AviAvni** [comment]
also it's interesting that for this code
```
printfn "%A" None
printfn "%A" ("", None)
```
the output is
![image](https://user-images.githubusercontent.com/1097246/36149044-e14c1f88-10c6-11e8-80c9-95583297614d.png)

**Comment by ChrSteinert** [comment]
But
```
> printfn "%O" None
- printfn "%O" ("", None);;
<null>
(, )
```

**Comment by AviAvni** [comment]
the <null> is comming from https://github.com/Microsoft/visualfsharp/pull/1776/files#diff-f42bdc58d3a07dc6e26affa6147f5759R813

**Comment by matthid** [comment]
```fsharp
> let v = None : int option;;
val v : int option = None

> v.Value;;
System.NullReferenceException: Object reference not set to an instance of an object.
   at <StartupCode$FSI_0005>.$FSI_0005.main@()
Stopped due to error
> v.IsSome;;
val it : bool = false

> v.IsNone;;
val it : bool = true
```

At the very least it is a bit surprising. Question is how does the compiler make `IsNone` and `IsSome` work? Should the compiler do the same for other members if one of the fields has `UseNullAsTrueValue`?

The `.Value` has some attribute applied it seems:

https://github.com/Microsoft/visualfsharp/blob/44fa027b308681a1b78a089e44fa1ab35ff77b41/src/fsharp/FSharp.Core/prim-types.fsi#L1594-L1640

https://github.com/Microsoft/visualfsharp/blob/44fa027b308681a1b78a089e44fa1ab35ff77b41/src/fsharp/FSharp.Core/prim-types.fs#L2946-L2976

**Comment by AviAvni** [comment]
@matthid None.IsNone compile to 
```
IL_0000: ldnull
IL_0001: call bool class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<object>::get_IsNone(class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<!0>)
```

so it doesn't call the instance method it calls the method from Option module
this help maybe we can do it also for ToString

**Comment by matthid** [comment]
I don't think it calls the option module. It rather compiles the members as static.

**Comment by dsyme** [comment]
This is by design because (for better or worse) option values are a reference type using `null` as a representation for None. 

We could emit a warning when `x.ToString()`, `x.GetHashCode` or `x.Equals(y)` are used directly on option-typed values, suggesting to use ``string`` or ``x = y`` instead. Or more generally when `Equals`, `GetHashCode` or `ToString` are used on F#-declared types known to have null as a representation.  I'll change the title of the bug to reflect this.

**Comment by BentTranberg** [comment]
As @AviAvni says above, the problem is also with unit, and I'm thinking perhaps that's worse than the rest of the problems described, but I'm not at all sure how much more "for worse", if at all. I just now ran into the issue while experimenting with unit as a kind of "noop", and experienced something like this ...

```
let myFunc a = System.Console.WriteLine (a.GetType().Name)

myFunc 1   // Int32
myFunc (1, ())   // Tuple`2
myFunc ()   // exception: Object reference not set to an instance of an object, in myFunc

```

**Comment by matthid** [comment]
@BentTranberg I don't find your case too concerning. 

`myFunc` is generic after all, therefore it needs to handle `null` in any case. For example `myFunc null` or `myFunc (null:string)`.

(And the "proper" way to write the function is:
`let myFunc (a:'a) = System.Console.WriteLine (typeof<'a>.Name)`)

But it is in fact a bit "unusal" as you cannot use `isNull` as you would assume to "check for `null`"

**Comment by BentTranberg** [comment]
Sorry, off-topic, but @matthid, you're wrong about the correct way to write the function. What you're doing is getting the compile time type, while my function gets the runtime type, so there is no right or wrong.

```
type MyBaseType() = class end
type MyDerivedType() = inherit MyBaseType()
let getRuntimeType a = System.Console.WriteLine ("Runtime type is " + (a.GetType().Name))
let getCompileTimeType (a: 'a) = System.Console.WriteLine ("Compile time type is " + (typeof<'a>.Name))
let myBaseRef1: MyBaseType = new MyBaseType()
let myBaseRef2: MyBaseType = new MyDerivedType() :> MyBaseType
getRuntimeType myBaseRef1 // Runtime type is MyBaseType
getRuntimeType myBaseRef2 // Runtime type is MyDerivedType
getCompileTimeType myBaseRef1 // Compile time type is MyBaseType
getCompileTimeType myBaseRef2 // Compile time type is MyBaseType

```

**Comment by matthid** [comment]
@BentTranberg Yes sure sorry, my point still holds.

**Comment by cartermp** [comment]
I would say that this would be a better candidate for an analyzer than a warning. cc @Krzysztof-Cieslak

**Comment by Krzysztof-Cieslak** [comment]
@cartermp, well, yes it can definitely be implemented as an analyzer. But we need to remember that those are 
1. not officially supported. 
2. they're FSAC-only at the moment (so no VS or Rider)
